File: MLP_Classification.py
#!/usr/bin/python

# Multi Layer Perceptron - Classification with gradient
# decent and backpropogation for weights update using
# Sigmoidal activation function.

import numpy as np
from random import randint
from math import sqrt

# Load file containing training data set.
# Output must be in class coded label format.
NTrain = np.loadtxt('xyz.tra',dtype = float)
m,n = NTrain.shape
NTD = m

# Input parameters needed to be initialized.
# inp = no. of input neurons i.e. input features/dimensions.
# hid = no. of hidden neurons.
# out = no. of output neurons i.e. no. of output classes.
# lam = learning rate.
# epo = epoch cycles.
inp = 19
hid = 100
out = 7
lam = 1.e-02
epo = 1000

# input and output weight matrices initialized at random.
Wi = 0.001*(np.random.rand(hid,inp)*2.0-1.0)
Wo = 0.001*(np.random.rand(out,hid)*2.0-1.0)

# Train the network.
for ep in range(epo):
	sumerr = 0
	miscla = 0
	for sa in range(NTD):
		xx = np.array([NTrain[sa,0:inp]]).T			# Current sample
		tt = np.array([NTrain[sa,inp:]]).T			# Current traget
		Yh = 1/(1 + np.exp(-Wi.dot(xx)))			# Hidden output
		Yo = Wo.dot(Yh)						# Predicted output
		er = tt - Yo						# Error
		Wo = Wo + lam * (er.dot((Yh.T)))			# Update rule for output weights
		Wi = Wi + lam * (((Wo.T).dot(er))*Yh*(1-Yh))*(xx.T)	# Update rule for input weights
		sumerr = sumerr + np.sum(er**2)
		ca = np.where(tt==1)[0][0]				# Actual class
		cp = np.where(Yo==max(Yo))[0][0]			# Predicted class
		if ca!=cp:
			miscla += 1
	print [sumerr,miscla]

# Validate the network.
conftra = np.zeros(shape=(out,out))
res_tra = np.zeros(shape=(NTD,2))
for sa in range(NTD):
	xx = np.array([NTrain[sa,0:inp]]).T
	tt = np.array([NTrain[sa,inp:]]).T
	Yh = 1/(1 + np.exp(-Wi.dot(xx)))
	Yo = Wo.dot(Yh)
	ca = np.where(tt==1)[0][0]
	cp = np.where(Yo==max(Yo))[0][0]
	conftra[ca][cp] += 1
	res_tra[sa,:] = [ca,cp]
print "conftra:"
print conftra

# Test the network.
# Load file containing testing data set.
# Output must be in class lable format and not in class coded labels.
NFeature = np.loadtxt('xyz.tes',dtype = float)
m,n = NFeature.shape
NTD = m
conftes = np.zeros(shape=(out,out))
res_tes = np.zeros(shape=(NTD,2))
for sa in range(NTD):
	xx = np.array([NFeature[sa,0:inp]]).T
	ca = np.array([NFeature[sa,inp:]])
	Yh = 1/(1 + np.exp(-Wi.dot(xx)))
	Yo = Wo.dot(Yh)
	cp = np.where(Yo==max(Yo))[0][0]
	conftes[int(ca-1)][cp] += 1
	res_tes[sa,:] = [ca,cp]
print "conftes:"
print conftes

File: MLP_Classification.m
% Program for Risk sensitive MLP..........................................

clear all
close all
clc

% Load the training data..................................................
Ntrain=load('xyz.tra');
[NTD,~] = size(Ntrain);

% Initialize the Algorithm Parameters.....................................
inp = 19;          % No. of input neurons
hid = 100 ;        % No. of hidden neurons
out = 7;            % No. of Output Neurons
lam = 1.e-02;       % Learning rate
epo = 1000;

% Initialize the weights..................................................
Wi = 0.001*(rand(hid,inp)*2.0-1.0);  % Input weights
Wo = 0.001*(rand(out,hid)*2.0-1.0);  % Output weights


% Train the network.......................................................
for ep = 1 : epo
    sumerr = 0;
    miscla = 0;
    for sa = 1 : NTD
        xx = Ntrain(sa,1:inp)';     % Current Sample
        tt = Ntrain(sa,inp+1:end)'; % Current Target
        Yh = 1./(1+exp(-Wi*xx));    % Hidden output
        Yo = Wo*Yh;                 % Predicted output
        er = tt - Yo;               % Error
        Wo = Wo + lam * (er * Yh'); % update rule for output weight
        Wi = Wi + lam * ((Wo'*er).*Yh.*(1-Yh))*xx';    %update for input weight
        sumerr = sumerr + sum(er.^2);
        ca = find(tt==1);           % actual class
        [~,cp] = max(Yo);           % Predicted class
        if ca~=cp 
            miscla = miscla + 1;
        end
    end
    disp([sumerr miscla])
%     save -ascii Wi.dat Wi;
%     save -ascii Wo.dat Wo;
end

% Validate the network.....................................................
conftra = zeros(out,out);
res_tra = zeros(NTD,2);
for sa = 1: NTD
        xx = Ntrain(sa,1:inp)';     % Current Sample
        tt = Ntrain(sa,inp+1:end)'; % Current Target
        Yh = 1./(1+exp(-Wi*xx));    % Hidden output
        Yo = Wo*Yh;                 % Predicted output
        ca = find(tt==1);           % actual class
        [~,cp] = max(Yo);           % Predicted class
        conftra(ca,cp) = conftra(ca,cp) + 1;
        res_tra(sa,:) = [ca cp];
end
disp(conftra)

% Test the network.........................................................
NFeature=load('xyz.tes');
[NTD,~]=size(NFeature);
conftes = zeros(out,out);
res_tes = zeros(NTD,2);
for sa = 1: NTD
        xx = NFeature(sa,1:inp)';   % Current Sample
        ca = NFeature(sa,end);      % Actual class
        Yh = 1./(1+exp(-Wi*xx));    % Hidden output
        Yo = Wo*Yh;                 % Predicted output
        [~,cp] = max(Yo);           % Predicted class
        conftes(ca,cp) = conftes(ca,cp) + 1;
        res_tes(sa,:) = [ca cp];
end
disp(conftes)
